### PR TITLE
Parallel

### PR DESCRIPTION
**Pull request by jeremylorino**
Working towards resolution of #16 will reference #4 because there was conversation regarding this in the issue.

### CODE REVIEW

**Comment by jeremylorino** [comment]
@yoiang check it. this should solve #16 with CLI params included.

also brings codacy to almost 100% across the board. all issues squashed except one.

**Comment by yoiang** [comment]
The only thing I'd say is it would be great to have `bluebird` as an optional dependency rather than required and have the backup code's contents largely agnostic which still taking advantage of `bluebird`'s or any future Promise libraries particulars. This may be having my cake and eating it too ;)

I apologize, I didn't thank you for this work earlier up, this is great! 😄

**Comment by jeremylorino** [comment]
@yoiang how about this

removed bluebird as a dependency and localized the use of `Promise` inside `FirestoreBackup` and allows it to be overridden after initialization.

we can add bluebird back after this and use it if we want.

File: lib/firestore.js [hunk 1]
```diff
@@ -169,11 +184,12 @@ export class FirestoreBackup {
       throw new Error('Unable to create backup path for Collection \'' + collection.id + '\': ' + error)
     }
 
-    return Promise.resolve(collection.get())
+    return this.Promise.resolve(collection.get())
       .then((documentSnapshots) => documentSnapshots.docs)
       .map((document) => {
         return this.backupDocument(document, backupPath + '/' + document.id, logPath + collection.id + '/')
       }, { concurrency: this.options.requestCountLimit })
+      .catch(err=>console.error(err))
```
Comment (yoiang) [inline lib/firestore.js]: I was wondering why we aren't bubbling up our errors anymore in the Promise's return result

Comment (yoiang) [reply lib/firestore.js]: Ah my mistake, you removed those in the latest commit!

File: lib/firestore.js [hunk 2]
```diff
@@ -125,12 +124,28 @@ export class FirestoreBackup {
...
+  get Promise() {
+    return this._PromiseLibrary
+  }
+
+  /**
+   * The Promise library used to process asynchronous functions.
+   */
+  set Promise(value: Object) {
+    // TODO: validate this ia valid Promise-like object
+    this._PromiseLibrary = addMapFunction(value)
```
Comment (yoiang) [inline lib/firestore.js]: Why are we mapping through the promise serial resolver again vs doing them all async? Does this tie into your analysis work?

Comment (jeremylorino) [reply lib/firestore.js]: It was a path I thought I wanted to go down until I decided to just write a func to handle promises at a given concurrency. See the latest commits.

**Comment by yoiang** [comment]
It looks great, thank you again  :)]

**Pull request merged**